File: Implement_By_Python/RayTracing/cpp_impl/src/base_object.cpp
# include "base_object.h"
# include <glm/glm.hpp>

BaseObject::BaseObject(
    const glm::vec3& position,
    const glm::vec3& color, 
    const float reflection,
    const float diffuse, 
    const float specular_coef,
    const float specular_k
) : position(position), color(color), reflection(reflection), diffuse(diffuse), 
    specular_coef(specular_coef), specular_k(specular_k) {}

float BaseObject::get_diffuse() const {
    return diffuse;
}

glm::vec3 BaseObject::get_color() const {
    return color;
}

float BaseObject::get_specular_coef() const {
    return specular_coef;
}

float BaseObject::get_specular_k() const {
    return specular_k;
}

float BaseObject::get_reflection() const {
    return reflection;
}

File: refactor_version/RayTracing/py_impl/base_object.py
import numpy as np

class BaseObject:
    def __init__(
            self,
            position: list,
            color: list,
            reflection: float,
            diffuse: float,
            specular_coef: float,
            specular_k: float
        ) -> None:

        self.position = np.array(position)
        self.color = np.array(color)
        self.reflection = reflection
        self.diffuse = diffuse
        self.specular_coef = specular_coef
        self.specular_k = specular_k

    def intersect(self):
        raise NotImplementedError
    
    def get_normal(self):
        raise NotImplementedError